libs/git/clone.go

```diff
@@ -65,6 +68,19 @@ func Clone(ctx context.Context, url, reference, targetPath string) error {
 
 	// wait for git clone to complete
 	err = cmd.Wait()
+	// Git repos hosted via HTTP do not support shallow cloning. We try with
+	// a deep clone this time
+	if err != nil && strings.Contains(cmdErr.String(), "dumb http transport does not support shallow capabilities") {
+		retryCmd := exec.CommandContext(ctx, "git", opts.args(false)...)
+		var retryCmdErr bytes.Buffer
+		retryCmd.Stderr = &retryCmdErr
+		retryErr := retryCmd.Run()
+		if retryErr != nil {
+			return fmt.Errorf("git clone failed: %w. %s", retryErr, retryCmdErr.String())
+		} else {
+			return nil
+		}
```

Comment:
This logic is duplicated now.

Idea: add an unexported function `clone` that takes only an `opts` argument and runs the command like this function already does. Then you can call that from this function the first time around with `Shallow: true`, and if it returns said error, flip the shallow field, and try again. Then we don't need to keep the duplicate run code.

Reply:
Good point. Done.